pkg/proxy/iptables/proxier.go

```diff
@@ -572,45 +590,56 @@ func (proxier *Proxier) OnEndpointsSynced() {
 	proxier.syncProxyRules(syncReasonEndpoints)
 }
 
-// Convert a slice of api.Endpoints objects into a map of service-port -> endpoints.
-func buildNewEndpointsMap(allEndpoints endpointsMap, curMap proxyEndpointMap, hostname string) (newMap proxyEndpointMap, hcEndpoints map[types.NamespacedName]int, staleSet map[endpointServicePair]bool) {
+func (proxier *Proxier) mergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		proxier.endpointsMap[svcPort] = endpointsMap[svcPort]
+	}
 
-	// return values
-	newMap = make(proxyEndpointMap)
-	hcEndpoints = make(map[types.NamespacedName]int)
-	staleSet = make(map[endpointServicePair]bool)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn, ips := range localIPs {
+		proxier.hcEndpoints[nsn] = len(ips)
+	}
+}
 
-	// Update endpoints for services.
-	for _, endpoints := range allEndpoints {
-		accumulateEndpointsMap(endpoints, hostname, &newMap)
+func (proxier *Proxier) unmergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		delete(proxier.endpointsMap, svcPort)
 	}
-	// Check stale connections against endpoints missing from the update.
-	// TODO: we should really only mark a connection stale if the proto was UDP
-	// and the (ip, port, proto) was removed from the endpoints.
-	for svcPort, epList := range curMap {
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn := range localIPs {
+		delete(proxier.hcEndpoints, nsn)
+	}
+}
+
+func (proxier *Proxier) detectStaleConnections(oldEndpointsMap, newEndpointsMap proxyEndpointsMap) {
```

Comment:
as above, I'd feel better with this as a non-method for easy testing.  I'd also rather not cache this state (I guess it just feels dirty) but it needs old and new state to delta.

What if, just an idea, we stored `proxier.curEndpoints` and `proxier.newEndpoints` - accumulate events into newEndpoints and call sync.  If sync successfully runs (after throttle added, even more important) then it moves new -> cur, and resets new.